File: AuthService/libsensateiot/data/measurementvalidator.cpp
/*
 * JSON measurement validator implementation.
 *
 * @author Michel Megens
 * @email  michel@michelmegens.net
 */

#include <sensateiot/data/measurementvalidator.h>

#include <rapidjson/document.h>
#include <rapidjson/rapidjson.h>

#include "parser.h"

namespace sensateiot::data
{
	std::pair<bool, models::RawMeasurement> MeasurementValidator::operator()(const std::string &str) const
	{
		try {
			rapidjson::Document json;
			json.Parse(str.c_str());

			return detail::ParseSingleMeasurement(json);
		} catch(std::exception &) {
			return std::make_pair(false, models::RawMeasurement());
		}
	}
}

File: SensateIoT.Platform.Router.Common/Validators/MeasurementValidator.cs
/*
 * Validate messages.
 *
 * @author Michel Megens
 * @email  michel@michelmegens.net
 */

using SensateIoT.Platform.Router.Data.DTO;

namespace SensateIoT.Platform.Router.Common.Validators
{
	public static class MeasurementValidator
	{
		private const int DataPointCountMax = 25;

		private const int MaxLat = 90;
		private const int MinLat = -90;
		private const int MaxLon = 180;
		private const int MinLon = -180;

		public static bool Validate(Measurement measurement)
		{
			if(measurement.Data.Count > DataPointCountMax) {
				return false;
			}

			if(measurement.Latitude == 0M && measurement.Longitude == 0M) {
				return true;
			}

			return measurement.Latitude >= MinLat && measurement.Latitude <= MaxLat && measurement.Longitude >= MinLon && measurement.Longitude <= MaxLon;
		}
	}
}
